azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -78,91 +71,59 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
             role_config = find(role_instance, "Configuration")
             self.role_config_name = findtext(role_config, "ConfigName")
             container = find(xml_doc, "Container")
             self.container_id = findtext(container, "ContainerId")
-            lbprobe_ports = find(xml_doc, "LBProbePorts")
-            self.load_balancer_probe_port = findtext(lbprobe_ports, "Port")
 
             AgentGlobals.update_container_id(self.container_id)
 
-            if full_goal_state:
-                reason = 'force update'
-            elif base_incarnation not in (None, self.incarnation):
-                reason = 'new incarnation'
-            else:
-                self.hosting_env = None
-                self.shared_conf = None
-                self.certs = None
-                self.ext_conf = None
-                self.remote_access = None
-                return
+            # these properties are populated by fetch_full_goal_state()
+            self._hosting_env_uri = findtext(xml_doc, "HostingEnvironmentConfig")
+            self.hosting_env = None
+            self._shared_conf_uri = findtext(xml_doc, "SharedConfig")
+            self.shared_conf = None
+            self._certs_uri = findtext(xml_doc, "Certificates")
+            self.certs = None
+            self._ext_conf_uri = findtext(xml_doc, "ExtensionsConfig")
+            self.ext_conf = None
+            self._remote_access_uri = findtext(container, "RemoteAccessInfo")
+            self.remote_access = None
+
         except Exception as exception:
             # We don't log the error here since fetching the goal state is done every few seconds
             raise ProtocolError(msg="Error fetching goal state", inner=exception)
 
+    def fetch_full_goal_state(self, wire_client):
         try:
-            logger.info('Fetching new goal state [incarnation {0} ({1})]', self.incarnation, reason)
+            logger.info('Fetching goal state [incarnation {0}]', self.incarnation)
 
-            uri = findtext(xml_doc, "HostingEnvironmentConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+            xml_text = wire_client.fetch_config(self._hosting_env_uri, wire_client.get_header())
             self.hosting_env = HostingEnv(xml_text)
 
-            uri = findtext(xml_doc, "SharedConfig")
-            xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+            xml_text = wire_client.fetch_config(self._shared_conf_uri, wire_client.get_header())
             self.shared_conf = SharedConfig(xml_text)
 
-            uri = findtext(xml_doc, "Certificates")
-            if uri is None:
-                self.certs = None
-            else:
-                xml_text = wire_client.fetch_config(uri, wire_client.get_header_for_cert())
+            if self._certs_uri is not None:
+                xml_text = wire_client.fetch_config(self._certs_uri, wire_client.get_header_for_cert())
```

Comment:
w.r.t. my comment above, do we know why the old code had a guard on the `fetch_config` call specifically for Certificates? was it just that we had actually experienced a value of `None` for Certificates in the wild, and not yet for the others, or is there some fundamental difference between Certificates and the other uris, like Certificates being explicitly optional?

Reply:
The exiting code assumes some of those fields are mandatory and some optional. All this is underspecified at best and not spec'ed most of the time. I do not want to change any error handling in this code, at least not as part of the work for fast track,